src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -839,22 +834,17 @@ GetModifyConnections(Task *task, bool markCritical, bool noNewTransactions)
 													 NULL);
 
 		/*
-		 * If already in a transaction, disallow expanding set of remote
-		 * transactions. That prevents some forms of distributed deadlocks.
+		 * If we're expanding the nodes that participate in the distributed
+		 * transaction, conform to MultiShardCommitProtocol.
 		 */
-		if (noNewTransactions)
+		if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC &&
+			InCoordinatedTransaction() &&
+			XactModificationLevel == XACT_MODIFICATION_DATA)
```

Comment:
Why the `XactModificationLevel` check here? Also, isn't this going to trigger for transactions that only touch a single shard? That might be costly for some rails etc users that use transactions for trivial one statement writes.